web/src/server/api/routers/utilities.ts

```diff
@@ -81,17 +82,102 @@ const isValidAndSecureUrl = async (urlString: string): Promise<boolean> => {
         ipAddresses.addresses6.every((ip) => !isPrivateIp(ip)))
     );
   } catch (error) {
-    console.error("Invalid URL:", error);
+    logger.info("Invalid URL:", error);
+    return false;
+  }
+};
+
+// Define a Zod schema for pre-signed S3 URL validation, based on https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html; 04.12.24
+const s3UrlSchema = z.object({
+  hostname: z
+    .string()
+    .refine((hostname) => hostname.endsWith(".amazonaws.com"), {
+      message: "Invalid hostname. Must be an S3 URL.",
+    }),
+  path: z.string(),
+  query: z.object({
+    "X-Amz-Algorithm": z.string().optional(),
+    "X-Amz-Credential": z.string().optional(),
+    "X-Amz-Date": z.string().optional(),
+    "X-Amz-Expires": z.string().optional(),
+    "X-Amz-SignedHeaders": z.string().optional(),
+    "X-Amz-Signature": z.string(),
+    "X-Amz-Security-Token": z.string().optional(),
+    "x-id": z.string().optional(), // Optional ID parameter, not documented in AWS docs as it is SDK specific, https://github.com/aws/aws-sdk-go-v2/blob/04e7aca073a0a7ed479aa37cad88a1cf58a979a1/service/s3/internal/customizations/presign_test.go#L35-L41
+  }),
+});
```

Comment:
Do we also have to consider presigned URLs for Azure, GCP, ... here? Those may be relevant for self-hosters.

Reply:
Not yet, I'd need to check if our current general check would catch these or not. Given no one has reported lack of support for Azure, GCP (or else) yet, I don't think we need to look into this right now :)